Styling refactor and modular components

yakkomajuri opened the pull request:
Taking significant steps towards improving our website codebase and facilitating component-based page creation. 

Key points:

### Moving away from global CSS styling to CSS modules as [recommended in the Gatsby Docs](https://www.gatsbyjs.com/tutorial/part-two/)

Currently we're importing CSS into JS files as if the styling was to apply to that page only. However, webpack bundles the CSS together into one file, meaning that our `pricing.css` file also affects our homepage and all other pages, for example.

By using CSS modules we can have truly component-scoped CSS that is much safer to use since we don't run the risk of breaking unrelated pages.

This PR moves `pricing.css` and `trial.css` to modules, but I didn't take on `index.css` and `features.css` yet. Those will naturally sort themselves out a bit as a function of the second goal of this PR.

### Component-driven Development

Right now, we're building every page from scratch, and "hard-coding" things like the Other Features block in the features pages. These should be **standalone re-usable components**. That will allow us to build pages more easily by just putting components together and passing the appropriate props.

#### Other Features

To start off with, I built the `OtherFeaturesBlock` component, which cut down a lot of unnecessary code. 

#### Page Header

I also created a separate component for the page header in features pages, which can still be further standardized. This will allow us to create consistent headers across the webpage with little to no styling and additional HTML. I also configured the main features page to also use this component.

##### Before:

<img width="1440" alt="Screenshot 2020-10-12 at 13 00 30" src="https://user-images.githubusercontent.com/38760734/95739786-aac5a880-0c8b-11eb-96d4-cf663f1553c2.png">

<img width="373" alt="Screenshot 2020-10-12 at 13 00 21" src="https://user-images.githubusercontent.com/38760734/95739732-971a4200-0c8b-11eb-8c64-552d817676d0.png">

##### After (Standardized):

<img width="374" alt="Screenshot 2020-10-12 at 12 59 50" src="https://user-images.githubusercontent.com/38760734/95739898-d2b50c00-0c8b-11eb-93bb-2c6a95c6625b.png">

<img width="1440" alt="Screenshot 2020-10-12 at 13 00 39" src="https://user-images.githubusercontent.com/38760734/95739814-b2854d00-0c8b-11eb-8b63-e225ae019767.png">


## Summary

I was going to submit this as a WIP, but it technically does make things better already without making anything (that I've seen) worse. 

A thorough review is needed, especially of the Pricing and product features pages. 

This also fixes the problem from #508

mariusandra commented on src/components/OtherFeaturesBlock/OtherFeaturesBlock.js:
```diff
@@ -0,0 +1,41 @@
+    )
+}
+
+export default OtherFeaturesBlock
```
I'd advocate against using default exports and just do `export const OtherFeaturesBlock ...`. There are [various technical reasons](https://basarat.gitbook.io/typescript/main-1/defaultisbad) why it makes sense and in the main app we mostly only use named exports.

Even if other (legacy?) code in this codebase is using default exports, it's a good idea to move away from them.

mariusandra commented on src/components/OtherFeaturesBlock/features.js:
```diff
@@ -0,0 +1,47 @@
+    },
+]
+
+export default featurePagesData
```
same comment about default exports here

mariusandra commented on src/components/OtherFeaturesBlock/index.js:
```diff
@@ -0,0 +1,3 @@
+import OtherFeaturesBlock from './OtherFeaturesBlock'
+
+export default OtherFeaturesBlock
```
... and this file can be omitted altogether I think.

I'll skip these types of comments in other components.

mariusandra reviewed (changes requested):
Hey! Looks like a great refactor and I definitely agree with the spirit of this PR! More encapsulation/componentization will make the .com repo better!

I have some code nitpicks below and I found a UX bug:

![2020-10-12 13 26 37](https://user-images.githubusercontent.com/53387/95741465-98993980-0c8e-11eb-97b8-76ce9d02821a.gif)

I didn't test every page on every resolution, just this features block stood out immediately... so don't please consider this a thorough review.

Regarding CSS modules. I had recently a [similar discussion](https://github.com/PostHog/posthog/pull/1819#discussion_r503153864) with @Twixes regarding [Styled Components](https://styled-components.com/) in the command palette PR (https://github.com/PostHog/posthog/pull/1819). The advice against using styled components was that the CSS classes generated all had some random hash prefix added to them (e.g. `.footer-e5f35av`) to avoid namespace conflicts.

CSS modules seem to produce similar classes in the final output:

![image](https://user-images.githubusercontent.com/53387/95742321-1b6ec400-0c90-11eb-9f8c-d7d9056b94e6.png)

![image](https://user-images.githubusercontent.com/53387/95742579-81f3e200-0c90-11eb-9e44-d9e91a9bbecd.png)

The problem is that these class names can change at will (either after small changes to the CSS... or maybe even after every recompile? I'm not sure) and this basically breaks autocapture and actions, which rely on stable class names to do its magic. Thus I'd strongly advise against using them.

In my experience, if regular "styles.scss" files are properly manually namespaced, css class conflicts are rare. For example naming the top level `<div>` as `.other-features-block-component` (or similar for pages) and then just scoping everything inside the `.scss` file to it:

```scss
.other-features-block-component {
  .wrapper {}
  .features {}
  .twoFeatures {}
  ...
}
```

Thus I'd advocate for dropping CSS modules and just coming up with a clear convention to scope CSS classes to components/pages.

yakkomajuri commented:
Thanks @mariusandra! Everything makes perfect sense. Good point on the CSS modules, thanks for reviewing so early since this now saves me some time as I'll stop using them straight away. Regarding default exports and things, I was just keeping consistency, but can move away from that and delete the index.js files as well.

My only question is about the UX bug: is that on a monitor? I don't have access to one here so can you test our features pages on master for the same issue? Of course I may have introduced a bug but I did simply abstract the components away, not mess with the CSS

yakkomajuri commented:
Added .scss, working on the features table and will push.

Update on the UX bug, I'm quite sure it's live rather than a function of this PR. Don't have a monitor to test but we're setting the image position in pixels rather than as a % of the screen size.

mariusandra commented:
Hey! Just zoom out and it'll be the same as having a huge monitor.

yakkomajuri commented:
ha, of course!

yakkomajuri commented:
Ok so I removed redundant files, fixed the UX bug, got rid of default exports (on my own components for now), and updated the mobile product features.

## Before (Confusing)

<img width="397" alt="Screenshot 2020-10-12 at 15 55 39" src="https://user-images.githubusercontent.com/38760734/95754763-d05eac00-0ca3-11eb-96e2-f67e63581b41.png">

## After 

<img width="397" alt="Screenshot 2020-10-12 at 15 55 19" src="https://user-images.githubusercontent.com/38760734/95754796-d94f7d80-0ca3-11eb-87a9-be9cabca390d.png">

yakkomajuri commented:
`product-features.js` is currently 649 lines in `master`. Here it is on this PR:

```js
import React from 'react'
import './styles/features.css'
import Layout from '../components/Layout'
import { PageHeader } from '../components/PageHeader'
import { FeaturesComparisonTable } from '../components/FeaturesComparisonTable'
import { FeaturesNav } from '../components/FeaturesNav'

function ProductFeatures() {
    return (
        <Layout>
            <PageHeader
                title="Features"
                tagline="We know you’re wondering how PostHog compares to other analytics tools, so we have made it easy
                for you to check out how feature-rich we actually are."
                styleKey="productFeatures"
                bgColor="navy"
            />
            <FeaturesNav />
            <FeaturesComparisonTable />
        </Layout>
    )
}

export default ProductFeatures
```

We should be developing like this.

yakkomajuri commented:
Converting to draft because I'd want to put in a few more components before merging. If someone really thinks this should be merged though feel free to go ahead.

mariusandra commented:
I wouldn't merge yet as the [/product-features page](https://deploy-preview-511--posthog.netlify.app/product-features) is still broken

![image](https://user-images.githubusercontent.com/53387/95836760-1f5a1f00-0d40-11eb-8579-f33fdb56908d.png)

yakkomajuri commented:
Huh, was working locally for me. Know the fix though. Will get to it later today + some commits I have locally atm

yakkomajuri commented:
@mariusandra so that's now fixed:

<img width="1440" alt="Screenshot 2020-10-13 at 23 03 06" src="https://user-images.githubusercontent.com/38760734/95921251-d7bbad80-0da8-11eb-89bc-a22f2b82a37a.png">

Some minor responsive issues on the homepage still which I'll go through tomorrow.

yakkomajuri commented:
This should now be good to go. There's a lot more refactoring I'm doing/want to do, but we should be splitting that across multiple PRs, otherwise this will grow, accumulate conflicts, and it will become impossible to merge.

@mariusandra any further comments on the code?

@jamesefhawkins how's the UX? The next step is to standardize all our features sections to use the same modular approach like we talked about. Snippet from homepage code for example:

```js
<Spacer />
<FeaturedSectionTextRight
    headerText="It all starts with event autocapture"
    listItem="01"
    descriptionText="PostHog autocaptures events and user behavior in your mobile or web app."
    image={retroAutocapture}
    color="yellow"
/>
<Spacer onlyDesktop={true} />
<FeaturedSectionTextLeft
    headerText="Understand how traffic really flows through your app"
    listItem="02"
    descriptionText="Know the pageviews and actions of every user in your app or on your website."
    image={trafficFlow}
    color="red"
    imgDesktopHeight="477px"
    imgDesktopWidth="50vw"
    imgStyle={{ backgroundSize: 'cover' }}
/>
<Spacer onlyDesktop={true} />
<FeaturedSectionTextRight
    headerText="Visualize product trends and retention"
    listItem="03"
    descriptionText="Powerful analytics to really understand what your users are doing and how to keep
    them coming back."
    image={visualizeTrends}
    color="navy"
/>
```

jamesefhawkins commented:
> This should now be good to go. There's a lot more refactoring I'm doing/want to do, but we should be splitting that across multiple PRs, otherwise this will grow, accumulate conflicts, and it will become impossible to merge.
> 
> @mariusandra any further comments on the code?
> 
> @jamesefhawkins how's the UX? The next step is to standardize all our features sections to use the same modular approach like we talked about. Snippet from homepage code for example:
> 
> ```js
> <Spacer />
> <FeaturedSectionTextRight
>     headerText="It all starts with event autocapture"
>     listItem="01"
>     descriptionText="PostHog autocaptures events and user behavior in your mobile or web app."
>     image={retroAutocapture}
>     color="yellow"
> />
> <Spacer onlyDesktop={true} />
> <FeaturedSectionTextLeft
>     headerText="Understand how traffic really flows through your app"
>     listItem="02"
>     descriptionText="Know the pageviews and actions of every user in your app or on your website."
>     image={trafficFlow}
>     color="red"
>     imgDesktopHeight="477px"
>     imgDesktopWidth="50vw"
>     imgStyle={{ backgroundSize: 'cover' }}
> />
> <Spacer onlyDesktop={true} />
> <FeaturedSectionTextRight
>     headerText="Visualize product trends and retention"
>     listItem="03"
>     descriptionText="Powerful analytics to really understand what your users are doing and how to keep
>     them coming back."
>     image={visualizeTrends}
>     color="navy"
> />
> ```

This is great. I think it *may* even be possible to connect this straight to figma to make this even easier ie https://interplayapp.com/

yakkomajuri commented:
Hmm interesting (re interplay). I'll take a look at that. Also will schedule a call with @lottiecoxon to talk about our design process and how that changes as a result of the refactoring I'm doing.

jamesefhawkins commented:
https://www.figma.com/blog/introducing-figma-to-react/ also looks interesting as an approach

mariusandra commented on .eslintrc.json:
```diff
@@ -15,6 +15,7 @@
         "react/prop-types": "off",
-        "no-const-assign": "off"
+        "no-const-assign": "off",
+        "react/jsx-key": "off"
```
This should be enabled again. Without a correct `key` it's totally random what will happen to arrays that are mapped into React components, etc. Components might just not render, etc.

https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-key.md

> If you are not using JSX then you can disable this rule.

We are :)

mariusandra commented on package.json:
```diff
@@ -29,6 +29,7 @@
         "gatsby-plugin-posthog-analytics": "^0.0.1",
         "gatsby-plugin-react-helmet": "^3.1.7",
         "gatsby-plugin-remove-trailing-slashes": "^2.3.11",
+        "gatsby-plugin-sass": "^2.3.16",
```
I guess this is fine, though nowadays instead of "sass" people tend to use "postcss" with optionally some "scss" plugins. Not sure if anything to change here.

yakkomajuri replied:
![fine](https://user-images.githubusercontent.com/38760734/95991069-38d49700-0e24-11eb-9836-da59e4fa118f.jpeg)

mariusandra commented on src/components/Sections/FeaturedSectionTextRight/index.js:
```diff
@@ -0,0 +1,53 @@
+    descriptionText,
+    image,
+    color = 'yellow',
+    imgDesktopHeight = '413px',
+    imgDesktopWidth = '621px',
+    mobileSpacerHeight = '50px',
```
These could be integers (`imgDesktopHeight = 413`) instead of strings with `px` added to them.

mariusandra reviewed (commented):
I had a quick look. The eslint issue should be fixed, the rest are more like nits. 

Somewhat related, if modular components is the path forward, it could make sense to also use storybook: https://storybook.js.org/

yakkomajuri commented:
@jamesefhawkins I'll merge this - did anything seem broken on the UI/UX side?

jamesefhawkins commented:
> @jamesefhawkins I'll merge this - did anything seem broken on the UI/UX side?

Let me check first thing tomorrow AM

jamesefhawkins commented:
> Let me check first thing tomorrow AM

lgtm

one non blocking issue

<img width="747" alt="Screenshot 2020-10-15 at 09 53 42" src="https://user-images.githubusercontent.com/47497682/96100803-acca7a00-0ecc-11eb-93d4-5d43b116a77f.png">

yakkomajuri commented:
@jamesefhawkins what screen size for issue 2 (table centering)? Can't reproduce...
